---
title: "VCFArray: DelayedArray objects with on-disk/remote VCF backend"
author:
- name: Qian Liu
  affiliation: Roswell Park Comprehensive Cancer Center, Buffalo, NY
- name: Valerie Obenchain
  affiliation: Fred Hutchinson Cancer Research Center, Seattle, WA
- name: Martin Morgan
  affiliation: Roswell Park Comprehensive Cancer Center, Buffalo, NY
date: "last edit: 08/22/2018"
output:
    BiocStyle::html_document:
        toc: true
        toc_float: true
package: VCFArray
vignette: >
  %\VignetteIndexEntry{VCFArray: DelayedArray objects with on-disk/remote VCF backend}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r options, eval=TRUE, echo=FALSE}
options(showHeadLines=3)
options(showTailLines=3)
```
```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction 
[VCFArray][] is a _Bioconductor_ package that represents VCF files as
objects derived from the [DelayedArray][] package and `DelayedArray`
class. It converts data entries from VCF file into a
`DelayedArray`-derived data structure. The backend VCF file could
either be saved on-disk locally or remote as online resources. Data
entries that could be extracted include the fixed data fields (REF,
ALT, QUAL, FILTER), information field (e.g., AA, AF...), and the
individual format field (e.g., GT, DP...). The array data generated
from fixed/information fields are one-dimensional `VCFArray`, with the
dimension being the length of the variants. The array data generated
from individual `FORMAT` field are always returned with the first
dimension being `variants` and the second dimension being
`samples`. This feature is consistent with the assay data saved in
`SummarizedExperiment`, and makes the `VCFArray` package interoperable
with other established _Bioconductor_ data infrastructure.

[VCFArray]: https://bioconductor.org/packages/VCFArray
[DelayedArray]: https://bioconductor.org/packages/DelayedArray

# Installation

1. Download the package. 

```{r getPackage, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("VCFArray")
```
The development version is also available to download from Github. 
```{r getDevel, eval=FALSE}
BiocManager::install("Liubuntu/VCFArray")
```

2. Load the package into R session.
```{r Load, message=FALSE}
library(VCFArray)
```

# VCFArray

## VCFArray constructor

To construct a `VCFArray` object, 3 arguments are needed: `file`,
`index` and `name`. The `file` argument could take either a character
string (VCF file name), or `VcfFile` object, or a `RangedVcfStack`
object. `name` argument must be specified to indicate which data entry
we want to extract from the input file. It's case-sensitive, and must
be consistent with the names from VCF header file.  `index` argument
will only be used to indicate the file path of the index file if it
does not exist.

The exported function `availableNames` returns a list of available
data entries from a VCF file that could be converted to `VCFArray`
object.

```{r avail}
args(VCFArray)
fl <- system.file("extdata", "chr22.vcf.gz", package = "VariantAnnotation")
availableNames(fl)
```

Since the index file for our vcf file already exists, the `index`
argument would not be needed (which is the most common case). So we
can construct the `VCFArray` object for the `GT` data entry in the
provided VCF file with arguments of `file` and `name` only.

```{r constructor}
VCFArray(file = fl, name = "GT")
```

We can also construct a `VCFArray` object with the `file` argument
being a `VcfFile` object.

```{r constructor2}
vcf <- VariantAnnotation::VcfFile(fl)
## VCFArray(file = vcf, name = "DS")
```

The `file` argument could also take `RangedVcfStack` object. Note that
an ordinary `VcfStack` object without `Range` information could not
be used to construct a `VCFArray`.

```{r rgstack}
rgfile <- system.file("extdata", "rgstack.rda", package = "VCFArray")
rgstack <- readRDS(rgfile)
rgstack
```

Here we choose the `name = SB`, which returns a 3-dimensional
`VCFArray` object, with the first 2 dimensions correspond to variants
and samples respectively.

```{r constructor3}
availableNames(rgstack)$geno
VCFArray(rgstack, name = "SB")
```

As the vignette title suggest, the backend VCF file could also be
remote files. Here we included an example of representing VCF file of
chromosome 22 from the 1000 Genomes Project (Phase 3). Note that for a
remote VCF file, the `index` argument must be specified. Since this
VCF files is relatively big, and it takes longer time, we only show
the code here without evaluation.

```{r remote, eval=FALSE}
chr22url <- "ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr22.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz"
chr22url.tbi <- "ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr22.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz.tbi"
va <- VCFArray(chr22url, index =chr22url.tbi, name = "GT")
```

## VCFArray methods

`VCFArray` represents VCF files as `DelayedArray` instances. It has
methods like `dim`, `dimnames` defined, and it inherits array-like
operations and methods from `DelayedArray`, e.g., the subsetting
method of `[`.     
Note that for 1-dimensional `VCFArray` objects that are generated from
the fixed / information data field of VCF file, when using `[` for
subsetting, we should always use `drop = FALSE` to ensure `VCFArray`
object as returned value.

```{r VCFArray}



```
## Internals: VCFArraySeed 

# sessionInfo()

```{r}
sessionInfo()
```
