## **NEW FEATURE TODO** next release: variant filtering/statistical functions. 
```{r}
library(VariantAnnotation)
fl <- system.file("extdata", "chr22.vcf.gz", package="VariantAnnotation")
vcf <- VcfFile(fl)
seed <- VCFArraySeed(fl, name = "GT")
VCFArray(seed)
VCFArray(seed)[1:12, ]  ## simple operation degrades "VCFMatrix" into "DelayedMatrix". 

table(isIndel(readVcf(vcffile(seed))))
#> FALSE  TRUE 
#>  9970   406 
table(isInsertion(readVcf(vcffile(seed))))
#> FALSE  TRUE 
#> 10202   174 
table(isSNV(readVcf(vcffile(seed))))
#> FALSE  TRUE 
#>   407  9969 
```                         
	
## **DOING** feature request to support `List` `DelayedArray`. 

issue reported: https://github.com/Bioconductor/DelayedArray/issues/27
Herve said feasible but requires major development. Will work on them
after September.  After done, the ".readVcf\_for\_class" function need
to remove the "array()" before returning the result.

## **DOING** BUG REPORT: bug in document(). error returns for VCFArray(seed) after document(): 

-- issue identified as bug in `devtools`: https://github.com/r-lib/devtools/issues/1845
-- workaround now. export `DelayedUnaryOp` and `DelayedOp` classes from `DelayedArray` package (locally modified and installed. install.packages("/home/qian/Documents/Research/DelayedArray", repos=NULL)).

```{r}
seed <- VCFArraySeed(fl, name = "GL")
VCFArray(seed)
#> error in validObject(.Object) : invalid class “DelayedMatrix” object: 
#>      'x' must have exactly 2 dimensions
traceback()
#> 10: stop(msg, ": ", errors, domain = NA)
#> 9: validObject(.Object)
#> 8: initialize(value, ...)
#> 7: initialize(value, ...)
#> 6: new(...)
#> 5: new2(Class, seed = seed)
#> 4: new_DelayedArray(seed, Class = "VCFArray") at VCFArray-class.R#104
#> 3: DelayedArray(seed)
#> 2: DelayedArray(seed) at VCFArray-class.R#131
#> 1: VCFArray(seed)
```

## **TODO** BUG: in numeric calculation. For `DelayedArray`?
```{r, numeric}
fl <- system.file("extdata", "chr22.vcf.gz", package="VariantAnnotation")
ds <- VCFArray(fl, name = "DS")
log(ds+5)
colSums(ds)  ## error
ds[colSums(ds) > 2, ,drop = FALSE]  #> FAIL. FIXME. 
```
## **DOING** New feature request for `GenomicFiles`. 

Martin suggests to add method of scanVcfHeader::VcfStack, to return
the shared header info for all VcfFile objects inside the VcfStack
object.

## **DONE** BUG report: show method for >2 dimensional array.  -- fixed in DelayedArray 0.7.28. 
  
	bug report: https://github.com/Bioconductor/DelayedArray/issues/25   
```{r}
rgstack
seed <- VCFArraySeed(rgstack, name="SB")
index <- list(1:2, NULL, NULL)
extract_array(seed, index)
index <- list(seq(to=nrow(seed), length.out=2), NULL, NULL)
extract_array(seed, index)
VCFArray(seed)  ## ERROR
```

## **DONE** type conversion from `VcfFile`.   
```
<10376 x 5> VCFMatrix object of type "Float":
```
Could directly extract the "Type" from vcf header. But how to generate the Float(0)/String(0) in R?
different data type within VcfFile and R?? 

**VCF**    **R type**
Integer    integer
Float      numeric
Flag       character
String     character
Character  character

- "Number" special cases: 
A: The field has one value per alternate allele. 
R: The field has one value for each possible allele, including the reference.
G: The field has one value for each possible genotype. 
. (dot): The number of possible values varies, is unknown or unbounded. (list)

## **DONE** How about the other data inside VCF? 
- todo: use array(DNAStringSetList/IntegerList/CharacterList) to construct DelayedArray objects. 
- rowData(): REF,ALT, QUAL, FILTER, paramRangeID?
- info(AC, AF, AN, DB, DP, FS, QD, ...)   
- IntegerList? (info: CIEND...)
- DNAStringSet (fixed: REF)
- DNAStringSetList (fixed: ALT)
- fixed(): 
  CHROM, POS, ID. -- do not need specific columns. 
  REF, ALT, QUAL, FILTER  readVcf(fixed = name)

```{r}
fl <- system.file("extdata", "chr22.vcf.gz", package="VariantAnnotation")
vcf <- VcfFile(fl)
hdr <- scanVcfHeader(fl)

param <- ScanVcfParam(fixed="ALT", info=NA, geno=NA)
res <- readVcf(vcf, genome="hg19", param=param)
ans <- fixed(res)[["ALT"]]  #> DNAStringSetList -- 
ans <- array(ans@unlistData)  #> fixed. 

getClass("XStringSetList")
getClass("XStringSet")
```
## **DONE** BUG: `type,VCFArray`.  
to be used as example in `VCFArray-methods.R`. 
```{r, type}
fl <- system.file("extdata", "chr22.vcf.gz", package="VariantAnnotation")
va <- VCFArray(fl, name = "GT")
DelayedArray::type(va)
```

## **DONE** rename "availableNames" into "vcfFields" as method. 

Pull request created https://github.com/Bioconductor/VariantAnnotation/pull/11
Added `vcfFields` generic and methods for "character" and `VcfFile`.   
Pull request created GenomicFiles: https://github.com/Bioconductor/GenomicFiles/pull/1
Added vcfFields methods for vcfStack.

Once merged, will remove the `availableNames` functions from `VCFArray` in R/man/tests/vignette...

## Some exploration for `VcfStack` file. 
```
rgfile <- system.file("extdata", "rgstack.rda", package="VCFArray")
rgfile
rgstack <- readRDS(rgfile)
readVcfStack(rgstack)  #> success
readVcfStack(rgstack, param = ScanVcfParam(info = NA)) #> fails.
readVcfStack(rgstack, param = ScanVcfParam(fixed = NA)) #> fails.

rg <- rowRanges(rgstack)

readVcfStack(rgstack,
             i = rg,
             j = 1:2)
```

